tests/core/test_data.py

```diff
@@ -0,0 +1,153 @@
+"""Unit tests for core.data module."""
+import os
+import shutil
+import tempfile
+import unittest
+
+import relentless
+
+class test_Directory(unittest.TestCase):
+    """Unit tests for core.data.Directory."""
+    def test_init(self):
+        """Test basic creation and methods of Directory."""
+        f = tempfile.TemporaryDirectory()
+
+        cwd = os.getcwd()
+
+        #test creation with existing path
+        d = relentless.data.Directory(f.name)
+        self.assertEqual(d.path, f.name)
+        self.assertEqual(d._start, None)
+        #enter and exit
+        with d:
+            self.assertEqual(d._start, cwd)
+        self.assertEqual(d._start, None)
+
+        #test creation with non-existent path (absolute)
+        foo = f.name + '/foo'
+        d1 = relentless.data.Directory(foo)
+        self.assertEqual(d1.path, foo)
+        self.assertEqual(d1._start, None)
+        #enter and exit
+        with d1:
+            self.assertEqual(d1._start, cwd)
+        self.assertEqual(d1._start, None)
+
+        #test creation with non-existent path (recursive)
+        foobar = foo + '/bar/foobar'
+        d2 = relentless.data.Directory(foobar)
+        self.assertEqual(d2.path, foobar)
+        self.assertEqual(d2._start, None)
+        #enter and exit
+        with d2:
+            self.assertEqual(d2._start, cwd)
+        self.assertEqual(d2._start, None)
+
+        #test creation with invalid directory path
+        x = tempfile.NamedTemporaryFile(dir=f.name)
+        with self.assertRaises(OSError):
+            d3 = relentless.data.Directory(x.name)
+        x.close()
+
+        #test unsuccessfully changing directories (by redundancy)
+        with d:
+            self.assertEqual(d._start, cwd)
+            with d1:
+                self.assertEqual(d1._start, d.path)
+                with d1:
+                    self.assertEqual(d1._start, d.path) #_start is unchanged
+                self.assertEqual(d1._start, None)
+                with d2:
+                    self.assertEqual(d2._start, d.path)
+                self.assertEqual(d2._start, None)
+            self.assertEqual(d1._start, None)
+        self.assertEqual(d._start, None)
+
+        #test unsuccessful exit after successful enter
+        with d2:
+            self.assertEqual(d2._start, cwd)
+            with d1:
+                self.assertEqual(d1._start, d2.path)
+                shutil.rmtree(foobar)
+            self.assertEqual(d1._start, None)
+            self.assertEqual(os.getcwd(), d1.path) #no exit
+
+        f.cleanup()
+
+    def test_context(self):
+        """Test context methods for Directory."""
+        f = tempfile.TemporaryDirectory()
+
+        #create nested directory structure
+        d = relentless.data.Directory(f.name)
+        d1 = d.directory('foo')
+        d2 = d.directory('bar')
+        d3 = d1.directory('bar/foobar')
```

Comment:
To avoid needing to `os.path.join` these together, maybe create the directory like this instead:
```suggestion
        d3 = d2.directory('foobar')
```